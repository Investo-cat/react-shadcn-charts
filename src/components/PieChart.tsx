import {
  PieChart as RePieChart,
  Pie,
  Cell,
  ResponsiveContainer,
  Tooltip,
} from "recharts";

type Props = {
  data: { label: string; value: number }[];
  dataKey?: string;
  colors?: string[];
  height?: number;
  className?: string;
};

export function PieChart({
  data,
  dataKey = "value",
  colors = [
    "var(--color-1, #2563eb)",
    "var(--color-2, #16a34a)",
    "var(--color-3, #f59e0b)",
    "var(--color-4, #dc2626)",
    "var(--color-5, #9333ea)",
  ],
  className,
}: Props) {
  return (
    <div className={className}>
      <ResponsiveContainer width="100%" height="100%">
        <RePieChart>
          <Pie
            data={data}
            dataKey={dataKey}
            nameKey={"label"}
            cx="50%"
            cy="50%"
            outerRadius={100}
            label
          >
            {data.map((_, index) => (
              <Cell
                key={`cell-${index}`}
                fill={colors[index % colors.length]}
              />
            ))}
          </Pie>
          <Tooltip />
        </RePieChart>
      </ResponsiveContainer>
    </div>
  );
}
