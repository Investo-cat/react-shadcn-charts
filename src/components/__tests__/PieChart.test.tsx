import React from "react";
import { render } from "@testing-library/react";

import { PieChart } from "../PieChart";

const pieChartMockData = [
  { label: "Apples", value: 40 },
  { label: "Bananas", value: 30 },
  { label: "Cherries", value: 20 },
];

jest.mock("recharts", () => {
  const Original = jest.requireActual("recharts");
  return {
    ...Original,
    ResponsiveContainer: ({ children }: any) =>
      React.cloneElement(children, { width: 400, height: 400 }),
  };
});

describe("PieChart", () => {
  it("matches snapshot", () => {
    const { container } = render(
      <PieChart data={pieChartMockData} isAnimationActive={false} />
    );
    expect(container).toMatchSnapshot();
  });

  it("applies custom className", () => {
    render(
      <PieChart
        data={pieChartMockData}
        isAnimationActive={false}
        className="test-class"
      />
    );
    expect(document.querySelector(".test-class")).not.toBeNull();
  });

  it("renders correct number of slices", async () => {
    const { container } = render(
      <PieChart data={pieChartMockData} isAnimationActive={false} />
    );

    const slices = container.querySelectorAll(".recharts-sector");
    expect(slices.length).toBe(pieChartMockData.length);
  });
});
