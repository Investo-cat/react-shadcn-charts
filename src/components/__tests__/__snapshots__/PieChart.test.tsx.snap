// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`PieChart matches snapshot 1`] = `
<div>
  <div>
    <div
      class="recharts-wrapper"
      style="position: relative; cursor: default; width: 400px; height: 400px;"
    >
      <svg
        class="recharts-surface"
        cx="50%"
        cy="50%"
        height="400"
        style="width: 100%; height: 100%;"
        viewBox="0 0 400 400"
        width="400"
      >
        <title />
        <desc />
        <defs>
          <clippath
            id="recharts1-clip"
          >
            <rect
              height="390"
              width="390"
              x="5"
              y="5"
            />
          </clippath>
        </defs>
        <g
          class="recharts-layer recharts-pie"
          tabindex="0"
        >
          <g
            class="recharts-layer recharts-pie-sector"
            tabindex="-1"
          >
            <path
              class="recharts-sector"
              cx="200"
              cy="200"
              d="M 300,200
    A 100,100,0,
    0,0,
    106.03073792140917,165.79798566743312
  L 200,200 Z"
              fill="var(--color-1, #2563eb)"
              name="Apples"
              role="img"
              stroke="#fff"
              tabindex="-1"
            />
          </g>
          <g
            class="recharts-layer recharts-pie-sector"
            tabindex="-1"
          >
            <path
              class="recharts-sector"
              cx="200"
              cy="200"
              d="M 106.03073792140917,165.79798566743312
    A 100,100,0,
    0,0,
    217.364817766693,298.48077530122083
  L 200,200 Z"
              fill="var(--color-2, #16a34a)"
              name="Bananas"
              role="img"
              stroke="#fff"
              tabindex="-1"
            />
          </g>
          <g
            class="recharts-layer recharts-pie-sector"
            tabindex="-1"
          >
            <path
              class="recharts-sector"
              cx="200"
              cy="200"
              d="M 217.364817766693,298.48077530122083
    A 100,100,0,
    0,0,
    300,200.00000000000003
  L 200,200 Z"
              fill="var(--color-3, #f59e0b)"
              name="Cherries"
              role="img"
              stroke="#fff"
              tabindex="-1"
            />
          </g>
          <g
            class="recharts-layer recharts-pie-labels"
          >
            <g
              class="recharts-layer"
            >
              <path
                class="recharts-curve recharts-pie-label-line"
                cx="200"
                cy="200"
                d="M217.36481776669305,101.5192246987792L220.83778132003164,81.82306963853503"
                fill="none"
                name="Apples"
                stroke="var(--color-1, #2563eb)"
              />
              <text
                alignment-baseline="middle"
                class="recharts-text recharts-pie-label-text"
                cx="200"
                cy="200"
                fill="var(--color-1, #2563eb)"
                name="Apples"
                stroke="none"
                text-anchor="start"
                x="220.83778132003164"
                y="81.82306963853503"
              >
                <tspan
                  dy="0em"
                  x="220.83778132003164"
                >
                  40
                </tspan>
              </text>
            </g>
            <g
              class="recharts-layer"
            >
              <path
                class="recharts-curve recharts-pie-label-line"
                cx="200"
                cy="200"
                d="M123.39555568810219,264.27876096865396L108.07466682572263,277.1345131623847"
                fill="none"
                name="Bananas"
                stroke="var(--color-2, #16a34a)"
              />
              <text
                alignment-baseline="middle"
                class="recharts-text recharts-pie-label-text"
                cx="200"
                cy="200"
                fill="var(--color-2, #16a34a)"
                name="Bananas"
                stroke="none"
                text-anchor="end"
                x="108.07466682572263"
                y="277.1345131623847"
              >
                <tspan
                  dy="0em"
                  x="108.07466682572263"
                >
                  30
                </tspan>
              </text>
            </g>
            <g
              class="recharts-layer"
            >
              <path
                class="recharts-curve recharts-pie-label-line"
                cx="200"
                cy="200"
                d="M276.6044443118978,264.27876096865396L291.92533317427734,277.13451316238474"
                fill="none"
                name="Cherries"
                stroke="var(--color-3, #f59e0b)"
              />
              <text
                alignment-baseline="middle"
                class="recharts-text recharts-pie-label-text"
                cx="200"
                cy="200"
                fill="var(--color-3, #f59e0b)"
                name="Cherries"
                stroke="none"
                text-anchor="start"
                x="291.92533317427734"
                y="277.13451316238474"
              >
                <tspan
                  dy="0em"
                  x="291.92533317427734"
                >
                  20
                </tspan>
              </text>
            </g>
          </g>
        </g>
      </svg>
      <div
        class="recharts-tooltip-wrapper"
        style="visibility: hidden; pointer-events: none; position: absolute; top: 0px; left: 0px;"
        tabindex="-1"
      >
        <div
          class="recharts-default-tooltip"
          style="margin: 0px; padding: 10px; background-color: rgb(255, 255, 255); border: 1px solid rgb(204, 204, 204); white-space: nowrap;"
        >
          <p
            class="recharts-tooltip-label"
            style="margin: 0px;"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;
